cosmovisor/README.md

```diff
@@ -0,0 +1,147 @@
+along with any other needed files (maybe the cli client? maybe some dlls?). `current` is a symlink to the currently
+active folder (so `current/bin/$DAEMON_NAME` is the binary)
+
+Note: the `<name>` after `upgrades` is the URI-encoded name of the upgrade as specified in the upgrade module plan.
+
+Please note that `$DAEMON_HOME/cosmovisor` just stores the *binaries* and associated *program code*.
+The `cosmovisor` binary can be stored in any typical location (eg `/usr/local/bin`). The actual blockchain
+program will store it's data under `$GAIA_HOME` etc, which is independent of the `$DAEMON_HOME`. You can
+choose to export `GAIA_HOME=$DAEMON_HOME` and then end up with a configuation like the following, but this
+is left as a choice to the admin for best directory layout.
+
+```
+.gaiad
+├── config
+├── data
+└── cosmovisor
+```
+
+## Usage
+
+Basic Usage:
+
+* The admin is responsible for installing the `cosmovisor` and setting it as a eg. systemd service to auto-restart, along with proper environmental variables
+* The admin is responsible for installing the `genesis` folder manually
+* The `cosmovisor` will set the `current` link to point to `genesis` at first start (when no `current` link exists)
+* The admin is (generally) responsible for installing the `upgrades/<name>` folders manually
+* The `cosmovisor` handles switching over the binaries at the correct points, so the admin can prepare days in advance and relax at upgrade time
+
+Note that chains that wish to support upgrades may package up a genesis `cosmovisor` tar file with this info, just as they
+prepare the genesis binary tar file. In fact, they may offer a tar file will all upgrades up to current point for easy download
+for those who wish to sync a fullnode from start.
+
+The `DAEMON` specific code, like the tendermint config, the application db, syncing blocks, etc is done as normal.
+The same eg. `GAIA_HOME` directives and command-line flags work, just the binary name is different.
+
+## Upgradeable Binary Specification
+
+In the basic version, the `cosmovisor` will read the stdout log messages
+to determine when an upgrade is needed. We are considering more complex solutions
+via signaling of some sort, but starting with the simple design:
+
+* when an upgrade is needed the binary will print a line that matches this
+regular expression: `UPGRADE "(.*)" NEEDED at height (\d+):(.*)`.
+* the second match in the above regular expression can be a JSON object with
+a `binaries` key as described above
+
+The name (first regexp) will be used to select the new binary to run. If it is present,
+the current subprocess will be killed, `current` will be upgraded to the new directory, 
+and the new binary will be launched.
+
+**Question** should we just kill the `cosmovisor` after it does the updates?
+so it gets a clean restart and just runs the new binary (under `current`).
+it should be safe to restart (as a service).
```

Comment:
Is this something for the discussion? I would suggest to kill the `cosmovisor` process after the update. Or may be there can be a simple flag `DAEMON_KILL_AND_RESTART_AFTER_UPGRADE` to handle this. Not a blocker on this PR, we can address it via a separate PR if we have consensus around this.

Reply:
We can handle this after if needed.

Reply:
Cool, can we remove this from here and add a tracking issue? Having it here might be confusing.